src/slacko.ml

```diff
@@ -765,33 +765,22 @@ let groups_list ?exclude_archived session =
       (match d |> groups_list_obj_of_yojson with
         | Result.Ok x -> `Success x.groups
         | Result.Error x -> `ParseFailure x)
-    | #bot_error
     | #parsed_auth_error as res -> res
     | _ -> `Unknown_error
 
-exception No_matches
-exception No_unique_matches
-exception Lookup_failed
-
 (* look up the id of query from results provided by the listfn *)
 let lookupk session listfn filterfn k =
   match%lwt listfn session with
-  | `Success channels -> (match List.filter filterfn channels with
-    | [] -> Lwt.fail No_matches
-    | [x] -> Lwt.return @@ k x
-    | _ -> Lwt.fail No_unique_matches)
-  | _ -> Lwt.fail Lookup_failed
+  | #parsed_auth_error as e -> Lwt.return e
+  | `Success items -> Lwt.return @@ k @@ List.filter filterfn items
```

Comment:
I agree that we shouldn't add more cases like `` `Too_many`` and `` `Too_few``, since most of the time I tried for the polymorphic variants to map to errors that the Slack API actually returns, not to Slacko-internal judgements upon the return values. At least where possible.

I actually quite like the approach of merged `k` and `filterfn`, because that pulls out the match on the 1 element list into a function which is meant to know what to do with it. The only complaint I see here is that returning `notfound` whenever *anything* is wrong with the function might be masking something seriously wrong with the library, the Slack API or both. I would rather do `| [] -> Lwt.return notfound | _ -> (* throw some exception here *)` since this is a case where the user probably cannot continue in a reasonable way. I am proposing an exception here, since this is a failure case that doesn't need to be represented in the signature and the user should most likely not try to catch the exception, since it basically only says "there is a bug in Slack or Slacko".